test/test/multi_threaded_map.cpp

```diff
@@ -0,0 +1,313 @@
+                        REQUIRE(val < size);
+                        last = val;
+                        if (!itr.MoveNext())
+                        {
+                            break;
+                        }
+
+                        // MoveNext is the only synchronized operation that has a predictable side effect we can validate
+                        ++increments[i];
+                    }
+                    catch (hresult_error const&)
+                    {
+                        // There's no "get if" function, so concurrent increment past the end is always possible...
+                        REQUIRE(!itr.HasCurrent());
+                        break;
+                    }
+                }
+            });
+        }
+
+        for (auto& t : threads)
+        {
+            t.join();
+        }
+
+        REQUIRE(!itr.HasCurrent());
+
+        auto totalIncrements = std::accumulate(std::begin(increments), std::end(increments), 0);
+        REQUIRE(totalIncrements == (size - 1));
+    }
+
+    // HasCurrent / GetMany
+    {
+        auto itr = map.First();
+        unique_thread threads[2];
+        int totals[std::size(threads)] = {};
+        for (int i = 0; i < std::size(threads); ++i)
+        {
+            threads[i] = unique_thread([&itr, &totals, i]
+            {
+                IKeyValuePair<int, T> vals[10];
+                while (itr.HasCurrent())
+                {
+                    // Unlike 'Current', 'GetMany' _is_ atomic in regards to read+increment
+                    auto len = itr.GetMany(vals);
+                    totals[i] += std::accumulate(vals, vals + len, 0, [](int curr, auto const& next) { return curr + next.Key(); });
+                }
+            });
+        }
+
+        for (auto& t : threads)
+        {
+            t.join();
+        }
+
+        // sum(i = 1 -> N){i} = N * (N + 1) / 2
+        auto total = std::accumulate(std::begin(totals), std::end(totals), 0);
+        REQUIRE(total == (size * (size - 1) / 2));
+    }
+}
+
+template <typename T>
+static void test_multi_writer(IMap<int, T> const& map)
+{
+    // Large enough that several threads should be executing concurrently
+    static constexpr uint32_t size = 10000;
+    static constexpr size_t threadCount = 8;
+
+    // Insert
+    unique_thread threads[threadCount];
+    for (int i = 0; i < threadCount; ++i)
+    {
+        threads[i] = unique_thread([&map, i]
+        {
+            auto off = i * size;
+            for (int j = 0; j < size; ++j)
+            {
+                map.Insert(j + off, conditional_box<T>(j));
+            }
+        });
+    }
+
+    for (auto& t : threads)
+    {
+        t.join();
+    }
+
+    REQUIRE(map.Size() == (size * threadCount));
+
+    // sum(i = 1 -> N){i} = N * (N + 1) / 2
+    auto sum = std::accumulate(begin(map), end(map), 0, [](int curr, auto&& pair) { return curr + conditional_unbox(pair.Value()); });
+    REQUIRE(sum == ((threadCount * (size - 1) * size) / 2));
+}
```

Comment:
I suppose it could be possible that an element gets inserted, but into the wrong part of the map's binary tree. It's probably more likely that in debug mode, at least, this would trip an assert. But in release mode, it's conceivable this state could occur - moreso if an insertion happened while the tree was rebalancing itself.

It's probably rare, but it's also super-easy to iterate the map and verify that the keys are all strictly increasing, so consider adding that little check. Aside from that, these tests look solid.